<!DOCTYPE html>
<html >
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
      <title>dna_features_viewer.biotools</title>
    
      <link rel="stylesheet" href="../../_static/pygments.css">
      <link rel="stylesheet" href="../../_static/theme.css">
      <link rel="stylesheet" href="../../_static/sphinx_press_theme.css">
      
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>

      <!-- sphinx script_files -->
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>

      
      <script src="../../_static/theme-vendors.js"></script>
      <script src="../../_static/theme.js" defer></script>
    
      <link rel="shortcut icon" href="../../_static/favicon.ico"/>
  <link rel="index" title="Index" href="../../genindex.html" />
  <link rel="search" title="Search" href="../../search.html" /> 
  </head>

  <body>
    <div id="app" class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <router-link to="../../index.html" class="home-link">
    
      <span class="site-name">DnaFeaturesViewer</span>
    
  </router-link>

  <div class="links">
    <navlinks class="can-hide">



  
    <div class="nav-item">
      <a href="https://github.com/Edinburgh-Genome-Foundry/DnaFeaturesViewer"
        class="nav-link external">
          Github <outboundlink/>
      </a>
    </div>
  

    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            



  
    <div class="nav-item">
      <a href="https://github.com/Edinburgh-Genome-Foundry/DnaFeaturesViewer"
        class="nav-link external">
          Github <outboundlink/>
      </a>
    </div>
  

            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Quick search</span>
    <div class="searchformwrapper">
      <form class="search" action="../../search.html" method="get">
        <input type="text" name="q" />
        <input type="submit" value="Search" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="../../index.html#more-biology-software">DNA Features Viewer</a></span>
      </p>
      <ul class="">
        
          <li class="toctree-l1 "><a href="../../ref.html" class="reference internal ">API Reference</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="../../tips.html" class="reference internal ">Tips</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="../../examples.html" class="reference internal ">Examples</a>

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="../../index.html">Docs</a> &raquo;</li>
    
      <li><a href="../index.html">Module code</a> &raquo;</li>
    
    <li>dna_features_viewer.biotools</li>
  </ul>
  

  <ul class="page-nav">
</ul>
  
</div>
<hr>
          <div class="content" role="main">
            
  <h1>Source code for dna_features_viewer.biotools</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">textwrap</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="k">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="kn">from</span> <span class="nn">Bio.PDB.Polypeptide</span> <span class="k">import</span> <span class="n">aa1</span><span class="p">,</span> <span class="n">aa3</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>


<div class="viewcode-block" id="complement"><a class="viewcode-back" href="../../ref.html#dna_features_viewer.biotools.complement">[docs]</a><span class="k">def</span> <span class="nf">complement</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the complement of the DNA sequence.</span>

<span class="sd">    For instance ``complement(&quot;ATGCCG&quot;)`` returns ``&quot;TACGGC&quot;``.</span>

<span class="sd">    Uses BioPython for speed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">())</span></div>


<div class="viewcode-block" id="reverse_complement"><a class="viewcode-back" href="../../ref.html#dna_features_viewer.biotools.reverse_complement">[docs]</a><span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the reverse-complement of the DNA sequence.</span>

<span class="sd">    For instance ``complement(&quot;ATGCCG&quot;)`` returns ``&quot;GCCGTA&quot;``.</span>

<span class="sd">    Uses BioPython for speed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span></div>


<span class="n">aa_short_to_long_form_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">_aa1</span><span class="p">:</span> <span class="n">_aa3</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">_aa3</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">_aa1</span><span class="p">,</span> <span class="n">_aa3</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">aa1</span> <span class="o">+</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="n">aa3</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;*&quot;</span><span class="p">])</span>
<span class="p">}</span>


<div class="viewcode-block" id="translate"><a class="viewcode-back" href="../../ref.html#dna_features_viewer.biotools.translate">[docs]</a><span class="k">def</span> <span class="nf">translate</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">,</span> <span class="n">long_form</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Translate the DNA sequence into an amino-acids sequence MLKYQT...</span>

<span class="sd">    If long_form is true, a list of 3-letter amino acid representations</span>
<span class="sd">    is returned instead ([&#39;Ala&#39;, &#39;Ser&#39;, ...]).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">result</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
    <span class="k">if</span> <span class="n">long_form</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="p">[</span><span class="n">aa_short_to_long_form_dict</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">result</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="extract_graphical_translation"><a class="viewcode-back" href="../../ref.html#dna_features_viewer.biotools.extract_graphical_translation">[docs]</a><span class="k">def</span> <span class="nf">extract_graphical_translation</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">location</span><span class="p">,</span> <span class="n">long_form</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a string of the &quot;graphical&quot; translation of a sequence&#39;s subsegment.</span>

<span class="sd">    Here &quot;graphical&quot; means that the amino acid sequence is always given</span>
<span class="sd">    left-to-right, as it will appear under the sequence in the plot. This matters</span>
<span class="sd">    when the location is on the -1 strand. In this case, the amino-acids are</span>
<span class="sd">    determined by (part of) the reverse-complement of the sequence, however</span>
<span class="sd">    the sequence returned will be the mirror of the translated sequence, as</span>
<span class="sd">    this is the left-to-right order in which the codons corresponding to the</span>
<span class="sd">    amino-acids appear in the sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    sequence</span>
<span class="sd">      An &quot;ATGC&quot; string.</span>

<span class="sd">    location</span>
<span class="sd">      Either (start, end) or (start, end, strand), with strand in (0, 1, -1).</span>

<span class="sd">    long_form</span>
<span class="sd">      if True, a list of 3-letter amino acid representations is returned instead</span>
<span class="sd">      ([&#39;Ala&#39;, &#39;Ser&#39;, ...]).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">location</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span> <span class="o">=</span> <span class="n">location</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">location</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">subsequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">subsequence</span> <span class="o">=</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">subsequence</span><span class="p">)</span>
    <span class="n">translation</span> <span class="o">=</span> <span class="n">translate</span><span class="p">(</span><span class="n">subsequence</span><span class="p">,</span> <span class="n">long_form</span><span class="o">=</span><span class="n">long_form</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">translation</span> <span class="o">=</span> <span class="n">translation</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">translation</span></div>


<div class="viewcode-block" id="load_record"><a class="viewcode-back" href="../../ref.html#dna_features_viewer.biotools.load_record">[docs]</a><span class="k">def</span> <span class="nf">load_record</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Load a Genbank file &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="annotate_biopython_record"><a class="viewcode-back" href="../../ref.html#dna_features_viewer.biotools.annotate_biopython_record">[docs]</a><span class="k">def</span> <span class="nf">annotate_biopython_record</span><span class="p">(</span>
    <span class="n">seqrecord</span><span class="p">,</span>
    <span class="n">location</span><span class="o">=</span><span class="s2">&quot;full&quot;</span><span class="p">,</span>
    <span class="n">feature_type</span><span class="o">=</span><span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span>
    <span class="n">margin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="o">**</span><span class="n">qualifiers</span>
<span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Add a feature to a Biopython SeqRecord.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    seqrecord</span>
<span class="sd">      The biopython seqrecord to be annotated.</span>

<span class="sd">    location</span>
<span class="sd">      Either (start, end) or (start, end, strand). (strand defaults to +1)</span>

<span class="sd">    feature_type</span>
<span class="sd">      The type associated with the feature</span>

<span class="sd">    margin</span>
<span class="sd">      Number of extra bases added on each side of the given location.</span>

<span class="sd">    qualifiers</span>
<span class="sd">      Dictionnary that will be the Biopython feature&#39;s `qualifiers` attribute.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">location</span> <span class="o">==</span> <span class="s2">&quot;full&quot;</span><span class="p">:</span>
        <span class="n">location</span> <span class="o">=</span> <span class="p">(</span><span class="n">margin</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqrecord</span><span class="p">)</span> <span class="o">-</span> <span class="n">margin</span><span class="p">)</span>

    <span class="n">strand</span> <span class="o">=</span> <span class="n">location</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">location</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span> <span class="k">else</span> <span class="mi">1</span>
    <span class="n">seqrecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">SeqFeature</span><span class="p">(</span>
            <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">location</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">location</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">strand</span><span class="p">),</span>
            <span class="n">qualifiers</span><span class="o">=</span><span class="n">qualifiers</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="n">feature_type</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="find_narrowest_text_wrap"><a class="viewcode-back" href="../../ref.html#dna_features_viewer.biotools.find_narrowest_text_wrap">[docs]</a><span class="k">def</span> <span class="nf">find_narrowest_text_wrap</span><span class="p">(</span><span class="n">text</span><span class="p">,</span> <span class="n">max_line_length</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Wrap the text into a multi-line text minimizing the longest line length.</span>

<span class="sd">    This is done by first wrapping the text using max_line_length, then</span>
<span class="sd">    attempt new wraps by iteratively decreasing the line_length, as long as the</span>
<span class="sd">    number of lines stays the same as with max_line_length.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">narrowest_wrap</span> <span class="o">=</span> <span class="n">textwrap</span><span class="o">.</span><span class="n">wrap</span><span class="p">(</span><span class="n">text</span><span class="p">,</span> <span class="n">max_line_length</span><span class="p">)</span>
    <span class="n">narrowest_width</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">l</span><span class="p">)</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">narrowest_wrap</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">line_length</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_line_length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">wrap</span> <span class="o">=</span> <span class="n">textwrap</span><span class="o">.</span><span class="n">wrap</span><span class="p">(</span><span class="n">text</span><span class="p">,</span> <span class="n">line_length</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">wrap</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">narrowest_wrap</span><span class="p">):</span>
            <span class="n">width</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">l</span><span class="p">)</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">wrap</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">width</span> <span class="o">&lt;</span> <span class="n">narrowest_width</span><span class="p">:</span>
                <span class="n">narrowest_wrap</span> <span class="o">=</span> <span class="n">wrap</span>
                <span class="n">narrowest_width</span> <span class="o">=</span> <span class="n">width</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">narrowest_wrap</span><span class="p">)</span></div>
</pre></div>

          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2016, Edinburgh Genome Foundry.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0 with <a href="https://github.com/schettino72/sphinx_press_theme">Press Theme</a>.
</div>
            </div>
          </div>
      </page>
    </div>
    
    
  </body>
</html>