<!DOCTYPE html>
<html >
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
      <title>API Reference</title>
    
      <link rel="stylesheet" href="_static/pygments.css">
      <link rel="stylesheet" href="_static/theme.css">
      <link rel="stylesheet" href="_static/sphinx_press_theme.css">
      
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>

      <!-- sphinx script_files -->
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>

      
      <script src="_static/theme-vendors.js"></script>
      <script src="_static/theme.js" defer></script>
    
      <link rel="shortcut icon" href="_static/favicon.ico"/>
  <link rel="index" title="Index" href="genindex.html" />
  <link rel="search" title="Search" href="search.html" />
  <link rel="next" title="Tips" href="tips.html" />
  <link rel="prev" title="DNA Features Viewer" href="index.html" /> 
  </head>

  <body>
    <div id="app" class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <router-link to="index.html" class="home-link">
    
      <span class="site-name">DnaFeaturesViewer</span>
    
  </router-link>

  <div class="links">
    <navlinks class="can-hide">



  
    <div class="nav-item">
      <a href="https://github.com/Edinburgh-Genome-Foundry/DnaFeaturesViewer"
        class="nav-link external">
          Github <outboundlink/>
      </a>
    </div>
  

    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            



  
    <div class="nav-item">
      <a href="https://github.com/Edinburgh-Genome-Foundry/DnaFeaturesViewer"
        class="nav-link external">
          Github <outboundlink/>
      </a>
    </div>
  

            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Quick search</span>
    <div class="searchformwrapper">
      <form class="search" action="search.html" method="get">
        <input type="text" name="q" />
        <input type="submit" value="Search" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="index.html#more-biology-software">DNA Features Viewer</a></span>
      </p>
      <ul class="current">
        
          <li class="toctree-l1 current"><a href="#" class="reference internal current">API Reference</a>

            
              <ul>
                
                  <li class="toctree-l2"><a href="#biopythontranslator" class="reference internal">BiopythonTranslator</a></li>
                
                  <li class="toctree-l2"><a href="#graphicrecord" class="reference internal">GraphicRecord</a></li>
                
                  <li class="toctree-l2"><a href="#circulargraphicrecord" class="reference internal">CircularGraphicRecord</a></li>
                
                  <li class="toctree-l2"><a href="#graphicfeature" class="reference internal">GraphicFeature</a></li>
                
                  <li class="toctree-l2"><a href="#module-dna_features_viewer.biotools" class="reference internal">Biotools</a></li>
                
              </ul>
            
          </li>

        
          <li class="toctree-l1 "><a href="tips.html" class="reference internal ">Tips</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="examples.html" class="reference internal ">Examples</a>

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
    
    <li>API Reference</li>
  </ul>
  

  <ul class="page-nav">
  <li class="prev">
    <a href="index.html"
       title="previous chapter">← DNA Features Viewer</a>
  </li>
  <li class="next">
    <a href="tips.html"
       title="next chapter">Tips →</a>
  </li>
</ul>
  
</div>
<hr>
          <div class="content" role="main">
            
  <div class="section" id="api-reference">
<h1>API Reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h1>
<div class="section" id="biopythontranslator">
<h2>BiopythonTranslator<a class="headerlink" href="#biopythontranslator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dna_features_viewer.BiopythonTranslator">
<em class="property">class </em><code class="sig-prename descclassname">dna_features_viewer.</code><code class="sig-name descname">BiopythonTranslator</code><span class="sig-paren">(</span><em class="sig-param">features_filters=()</em>, <em class="sig-param">features_properties=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/BiopythonTranslator.html#BiopythonTranslator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.BiopythonTranslator" title="Permalink to this definition">¶</a></dt>
<dd><p>A translator from SeqRecords to dna_features_viewer GraphicRecord.</p>
<p>This can be subclassed to create custom “themes” (see the example
<code class="docutils literal notranslate"><span class="pre">custom_biopython_translator.py</span></code> in the docs).</p>
<p>This class is meant to be customized by subclassing and changing the
methods (<code class="docutils literal notranslate"><span class="pre">compute_feature_label</span></code>, etc.) and/or the attributes
(<code class="docutils literal notranslate"><span class="pre">default_feature_color</span></code> etc).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>features_filters</strong></p>
<blockquote>
<div><p>List of filters (some_biopython_feature) =&gt; True/False.
Only features passing all the filters are kept.
This only works if you haven’t redefined <code class="docutils literal notranslate"><span class="pre">compute_filtered_features</span></code></p>
</div></blockquote>
<p><strong>features_properties</strong></p>
<blockquote>
<div><p>A function (feature)=&gt; properties_dict</p>
</div></blockquote>
</dd>
</dl>
<p class="rubric">Attributes</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>default_feature_color = “#7245dc”</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>graphic_record_parameters</strong></p></td>
<td><p>Dictionnary containing keyword arguments that will be passed to the (Circular)GraphicRecord constructor</p></td>
</tr>
<tr class="row-odd"><td><p><strong>ignored_features_types</strong></p></td>
<td><p>A list or tuple of strings indicating all the feature types that should always be ignored (i.e. not included in the graphic record) by the translator</p></td>
</tr>
<tr class="row-even"><td><p><strong>max_label_length</strong></p></td>
<td><p>Number of characters above which the labels will be printed cut and ended with an ellipsis “…”. This is to prevent extra-long labels from polluting the whole plots.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>max_line_length</strong></p></td>
<td><p>Feature labels with a number of characters above this number will be wrapped on 2 or more lines.</p></td>
</tr>
<tr class="row-even"><td><p><strong>label_fields</strong></p></td>
<td><p>This list of strings provides the order in which the different attributes of a Genbank feature will be considered, when automatically determining the feature label. For instance if the list is [“label”, “source”, “locus_tag”] and a feature has no label but has a “source”, the “source” will be displayed in the plots.</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="dna_features_viewer.BiopythonTranslator.compute_feature_box_color">
<code class="sig-name descname">compute_feature_box_color</code><span class="sig-paren">(</span><em class="sig-param">feature</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/BiopythonTranslator.html#BiopythonTranslator.compute_feature_box_color"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.BiopythonTranslator.compute_feature_box_color" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute a box_color for this feature.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.BiopythonTranslator.compute_feature_box_linewidth">
<code class="sig-name descname">compute_feature_box_linewidth</code><span class="sig-paren">(</span><em class="sig-param">feature</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/BiopythonTranslator.html#BiopythonTranslator.compute_feature_box_linewidth"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.BiopythonTranslator.compute_feature_box_linewidth" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute a box_linewidth for this feature.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.BiopythonTranslator.compute_feature_color">
<code class="sig-name descname">compute_feature_color</code><span class="sig-paren">(</span><em class="sig-param">feature</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/BiopythonTranslator.html#BiopythonTranslator.compute_feature_color"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.BiopythonTranslator.compute_feature_color" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute a color for this feature.</p>
<p>If the feature has a <code class="docutils literal notranslate"><span class="pre">color</span></code> qualifier it will be used. Otherwise,
the classe’s <code class="docutils literal notranslate"><span class="pre">default_feature_color</span></code> is used.</p>
<p>To change the behaviour, create a subclass of <code class="docutils literal notranslate"><span class="pre">BiopythonTranslator</span></code>
and overwrite this method.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.BiopythonTranslator.compute_feature_fontdict">
<code class="sig-name descname">compute_feature_fontdict</code><span class="sig-paren">(</span><em class="sig-param">feature</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/BiopythonTranslator.html#BiopythonTranslator.compute_feature_fontdict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.BiopythonTranslator.compute_feature_fontdict" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute a font dict for this feature.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.BiopythonTranslator.compute_feature_html">
<code class="sig-name descname">compute_feature_html</code><span class="sig-paren">(</span><em class="sig-param">feature</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/BiopythonTranslator.html#BiopythonTranslator.compute_feature_html"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.BiopythonTranslator.compute_feature_html" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the ‘label’ of the feature.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.BiopythonTranslator.translate_feature">
<code class="sig-name descname">translate_feature</code><span class="sig-paren">(</span><em class="sig-param">feature</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/BiopythonTranslator.html#BiopythonTranslator.translate_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.BiopythonTranslator.translate_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate a Biopython feature into a Dna Features Viewer feature.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.BiopythonTranslator.translate_record">
<code class="sig-name descname">translate_record</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">record_class=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/BiopythonTranslator.html#BiopythonTranslator.translate_record"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.BiopythonTranslator.translate_record" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new GraphicRecord from a BioPython Record object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>record</strong></p>
<blockquote>
<div><p>A BioPython Record object or the path to a Genbank file.</p>
</div></blockquote>
<p><strong>record_class</strong></p>
<blockquote>
<div><p>The graphic record class to use, e.g. GraphicRecord (default) or
CircularGraphicRecord. Strings ‘circular’ and ‘linear’ can also be
provided.</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="graphicrecord">
<h2>GraphicRecord<a class="headerlink" href="#graphicrecord" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dna_features_viewer.GraphicRecord">
<em class="property">class </em><code class="sig-prename descclassname">dna_features_viewer.</code><code class="sig-name descname">GraphicRecord</code><span class="sig-paren">(</span><em class="sig-param">sequence_length=None</em>, <em class="sig-param">sequence=None</em>, <em class="sig-param">features=()</em>, <em class="sig-param">feature_level_height=1</em>, <em class="sig-param">first_index=0</em>, <em class="sig-param">plots_indexing='biopython'</em>, <em class="sig-param">labels_spacing=8</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicRecord/GraphicRecord.html#GraphicRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicRecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Set of Genetic Features of a same DNA sequence, to be plotted together.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence_length</strong></p>
<blockquote>
<div><p>Length of the DNA sequence, in number of nucleotides</p>
</div></blockquote>
<p><strong>features</strong></p>
<blockquote>
<div><p>list of GraphicalFeature objects.</p>
</div></blockquote>
<p><strong>feature_level_height</strong></p>
<blockquote>
<div><p>Width in inches of one “level” for feature arrows.</p>
</div></blockquote>
<p><strong>annotation_height</strong></p>
<blockquote>
<div><p>Width in inches of one “level” for feature annotations.</p>
</div></blockquote>
<p><strong>first_index</strong></p>
<blockquote>
<div><p>Indicates the first index to plot in case the sequence is actually a
subsequence of a larger one. For instance, if the Graphic record
represents the segment (400, 420) of a sequence, we will have
<code class="docutils literal notranslate"><span class="pre">first_index=400</span></code> and <code class="docutils literal notranslate"><span class="pre">sequence_length=20</span></code>.</p>
</div></blockquote>
<p><strong>plots_indexing</strong></p>
<blockquote>
<div><p>Indicates which standard to use to show nucleotide indices in the plots.
If ‘biopython’, the standard python indexing is used (starting at 0).
If ‘genbank’, the indexing follows the Genbank standard (starting at 1).</p>
</div></blockquote>
<p><strong>labels_spacing</strong></p>
<blockquote>
<div><p>Number of pixels that will “pad” every labels to force some horizontal
space between two labels or between a label and the borders of a feature.</p>
</div></blockquote>
</dd>
</dl>
<p class="rubric">Attributes</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 5%" />
<col style="width: 95%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>default_font_family</strong></p></td>
<td><p>Default font to use for a feature that doesn’t declare a font.  default_ruler_color Default ruler color to use when no color is given at plot() time.  default_box_color Default box color for non-inline annotations. If set to None, no boxes will be drawn unless the features declare a box_color. If “auto”, a color (clearer version of the feature’s color) will be computed, for all features also declaring their box_color as “auto”.</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.annotate_feature">
<code class="sig-name descname">annotate_feature</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">feature</em>, <em class="sig-param">level</em>, <em class="sig-param">inline=False</em>, <em class="sig-param">max_label_length=50</em>, <em class="sig-param">max_line_length=30</em>, <em class="sig-param">padding=0</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.annotate_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a Matplotlib Text with the feature’s label.</p>
<p>The x-coordinates of the text are determined by the feature’s
<cite>x_center</cite> while the y-coordinates are determined by the <cite>level</cite>.</p>
<p>The text is horizontally and vertically centered.</p>
<p>The Arrow points in the direction of the feature’s strand.
If the feature has no direction (strand==0), the returned patch will
simply be a rectangle.</p>
<p>The x-coordinates of the patch are determined by the feature’s
<cite>start</cite> and <cite>end</cite> while the y-coordinates are determined by the <cite>level</cite></p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.autoselect_label_color">
<code class="sig-name descname">autoselect_label_color</code><span class="sig-paren">(</span><em class="sig-param">background_color</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.autoselect_label_color" title="Permalink to this definition">¶</a></dt>
<dd><p>Autselect a color for the label font.</p>
<p>In the current method the label will be black on clear backrgounds,
and white on dark backgrounds.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.bokeh_feature_patch">
<code class="sig-name descname">bokeh_feature_patch</code><span class="sig-paren">(</span><em class="sig-param">start</em>, <em class="sig-param">end</em>, <em class="sig-param">strand</em>, <em class="sig-param">width=0.3</em>, <em class="sig-param">level=0</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.bokeh_feature_patch" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dict with points coordinates of a Bokeh Feature arrow.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>start, end, strand</strong></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.coordinates_in_plot">
<code class="sig-name descname">coordinates_in_plot</code><span class="sig-paren">(</span><em class="sig-param">x</em>, <em class="sig-param">level</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicRecord/GraphicRecord.html#GraphicRecord.coordinates_in_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicRecord.coordinates_in_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a sequence position and height level into a (x, y) position.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.determine_annotation_height">
<code class="sig-name descname">determine_annotation_height</code><span class="sig-paren">(</span><em class="sig-param">levels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicRecord/GraphicRecord.html#GraphicRecord.determine_annotation_height"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicRecord.determine_annotation_height" title="Permalink to this definition">¶</a></dt>
<dd><p>By default the ideal annotation level height is the same as the
feature_level_height.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.finalize_ax">
<code class="sig-name descname">finalize_ax</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">features_levels</em>, <em class="sig-param">annotations_max_level</em>, <em class="sig-param">auto_figure_height=False</em>, <em class="sig-param">ideal_yspan=None</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.finalize_ax" title="Permalink to this definition">¶</a></dt>
<dd><p>Redefine y-bounds and figure height.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.initialize_ax">
<code class="sig-name descname">initialize_ax</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">draw_line</em>, <em class="sig-param">with_ruler</em>, <em class="sig-param">ruler_color=None</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.initialize_ax" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the ax: remove axis, draw a horizontal line, etc.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>draw_line</strong></p>
<blockquote>
<div><p>True/False to draw the horizontal line or not.</p>
</div></blockquote>
<p><strong>with_ruler</strong></p>
<blockquote>
<div><p>True/False to draw the indices indicators along the line.</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.plot">
<code class="sig-name descname">plot</code><span class="sig-paren">(</span><em class="sig-param">ax=None</em>, <em class="sig-param">figure_width=8</em>, <em class="sig-param">draw_line=True</em>, <em class="sig-param">with_ruler=True</em>, <em class="sig-param">ruler_color=None</em>, <em class="sig-param">plot_sequence=False</em>, <em class="sig-param">annotate_inline=True</em>, <em class="sig-param">max_label_length=50</em>, <em class="sig-param">max_line_length=30</em>, <em class="sig-param">level_offset=0</em>, <em class="sig-param">x_lim=None</em>, <em class="sig-param">figure_height=None</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot all the features in the same Matplotlib ax</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ax</strong></p>
<blockquote>
<div><p>The Matplotlib ax on which to plot the graphic record. If None is
provided, a new figure and ax is generated, the ax is returned at
the end.</p>
</div></blockquote>
<p><strong>figure_width</strong></p>
<blockquote>
<div><p>Width of the figure (only if no ax was provided and a new figure is
created) in inches.</p>
</div></blockquote>
<p><strong>draw_line</strong></p>
<blockquote>
<div><p>If True, a base line representing the sequence will be drawn.</p>
</div></blockquote>
<p><strong>with_ruler</strong></p>
<blockquote>
<div><p>If true, the sequence indices will be indicated at regular intervals.</p>
</div></blockquote>
<p><strong>plot_sequence</strong></p>
<blockquote>
<div><p>If True and the graphic record has a “sequence” attribute set, the
sequence will be displayed below the base line.</p>
</div></blockquote>
<p><strong>annotate_inline</strong></p>
<blockquote>
<div><p>If true, some feature labels will be displayed inside their
corresponding feature if there is sufficient space.</p>
</div></blockquote>
<p><strong>level_offset</strong></p>
<blockquote>
<div><p>All features and annotations will be pushed up by “level_offset”. Can
be useful when plotting several sets of features successively on a
same ax.</p>
</div></blockquote>
<p><strong>x_lim</strong></p>
<blockquote>
<div><p>Horizontal axis limits to be set at the end.</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.plot_feature">
<code class="sig-name descname">plot_feature</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">feature</em>, <em class="sig-param">level</em>, <em class="sig-param">linewidth=1.0</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.plot_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an Arrow Matplotlib patch with the feature’s coordinates.</p>
<p>The Arrow points in the direction of the feature’s strand.
If the feature has no direction (strand==0), the returned patch will
simply be a rectangle.</p>
<p>The x-coordinates of the patch are determined by the feature’s
<cite>start</cite> and <cite>end</cite> while the y-coordinates are determined by the <cite>level</cite></p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.plot_sequence">
<code class="sig-name descname">plot_sequence</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">location=None</em>, <em class="sig-param">y_offset=1</em>, <em class="sig-param">fontdict=None</em>, <em class="sig-param">background=('#f7fbff'</em>, <em class="sig-param">'#fffcf0')</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.plot_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot a sequence of nucleotides at the bottom of the plot.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ax</strong></p>
<blockquote>
<div><p>Which axes the translation should be plotted to</p>
</div></blockquote>
<p><strong>location</strong></p>
<blockquote>
<div><p>location of the segment to translate, either (start, end) or
(start, end, strand)</p>
</div></blockquote>
<p><strong>y_offset</strong></p>
<blockquote>
<div><p>Number of text levels under the plot’s base line where to draw the
nucleotides. Should be 1 if the nucleotide sequence is to be plotted
directly under the main line.</p>
</div></blockquote>
<p><strong>fontdict</strong></p>
<blockquote>
<div><p>Matplotlib fontdict for the text, e.g.
<code class="docutils literal notranslate"><span class="pre">{'size':</span> <span class="pre">11,</span> <span class="pre">'weight':'bold'}</span></code></p>
</div></blockquote>
<p><strong>background</strong></p>
<blockquote>
<div><p>tuple (color1, color2) of alternate colors to plot behind each
nucleotide position to guide vision. Leave to None for no background.</p>
</div></blockquote>
<p><strong>translation</strong></p>
<blockquote>
<div><p>Sequence of amino acids either as a string <code class="docutils literal notranslate"><span class="pre">'MAKG...'</span></code> or as a list
<code class="docutils literal notranslate"><span class="pre">['Met',</span> <span class="pre">'Ala',</span> <span class="pre">...]</span></code></p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.plot_translation">
<code class="sig-name descname">plot_translation</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">location=None</em>, <em class="sig-param">y_offset=2</em>, <em class="sig-param">fontdict=None</em>, <em class="sig-param">background=('#f5fff0'</em>, <em class="sig-param">'#fff7fd')</em>, <em class="sig-param">translation=None</em>, <em class="sig-param">long_form_translation=True</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.plot_translation" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot a sequence of amino-acids at the bottom of the plot.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ax</strong></p>
<blockquote>
<div><p>Which axes the translation should be plotted to</p>
</div></blockquote>
<p><strong>location</strong></p>
<blockquote>
<div><p>location of the segment to translate (start, end)</p>
</div></blockquote>
<p><strong>y_offset</strong></p>
<blockquote>
<div><p>Number of text levels under the plot’s base line where to draw the
amino acid names. Should be 2 if the nucleotide sequence is also
plotted at level 1.</p>
</div></blockquote>
<p><strong>fontdict</strong></p>
<blockquote>
<div><p>Matplotlib fontdict for the text, e.g.
<code class="docutils literal notranslate"><span class="pre">{'size':</span> <span class="pre">11,</span> <span class="pre">'weight':'bold'}</span></code></p>
</div></blockquote>
<p><strong>background</strong></p>
<blockquote>
<div><p>tuple (color1, color2) of alternate colors to plot behind each
amino acid position to guide vision. Leave to None for no background.</p>
</div></blockquote>
<p><strong>translation</strong></p>
<blockquote>
<div><p>Sequence of amino acids either as a string <code class="docutils literal notranslate"><span class="pre">'MAKG...'</span></code> or as a list
<code class="docutils literal notranslate"><span class="pre">['Met',</span> <span class="pre">'Ala',</span> <span class="pre">...]</span></code></p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.plot_with_bokeh">
<code class="sig-name descname">plot_with_bokeh</code><span class="sig-paren">(</span><em class="sig-param">figure_width=5</em><span class="sig-paren">)</span><a class="headerlink" href="#dna_features_viewer.GraphicRecord.plot_with_bokeh" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the graphic record using Bokeh.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.span">
<em class="property">property </em><code class="sig-name descname">span</code><a class="headerlink" href="#dna_features_viewer.GraphicRecord.span" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the display span (start, end) accounting for first_index.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.split_overflowing_features_circularly">
<code class="sig-name descname">split_overflowing_features_circularly</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicRecord/GraphicRecord.html#GraphicRecord.split_overflowing_features_circularly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicRecord.split_overflowing_features_circularly" title="Permalink to this definition">¶</a></dt>
<dd><p>Split the features that overflow over the edge for circular
constructs (inplace).</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicRecord.to_biopython_record">
<code class="sig-name descname">to_biopython_record</code><span class="sig-paren">(</span><em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicRecord/GraphicRecord.html#GraphicRecord.to_biopython_record"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicRecord.to_biopython_record" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="circulargraphicrecord">
<h2>CircularGraphicRecord<a class="headerlink" href="#circulargraphicrecord" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dna_features_viewer.CircularGraphicRecord">
<em class="property">class </em><code class="sig-prename descclassname">dna_features_viewer.</code><code class="sig-name descname">CircularGraphicRecord</code><span class="sig-paren">(</span><em class="sig-param">sequence_length</em>, <em class="sig-param">features</em>, <em class="sig-param">top_position=0</em>, <em class="sig-param">feature_level_height=0.2</em>, <em class="sig-param">annotation_height='auto'</em>, <em class="sig-param">labels_spacing=12</em>, <em class="sig-param">**kw</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/CircularGraphicRecord/CircularGraphicRecord.html#CircularGraphicRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.CircularGraphicRecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Set of Genetic Features of a same DNA sequence, to be plotted together.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence_length</strong></p>
<blockquote>
<div><p>Length of the DNA sequence, in number of nucleotides</p>
</div></blockquote>
<p><strong>features</strong></p>
<blockquote>
<div><p>list of GraphicalFeature objects.</p>
</div></blockquote>
<p><strong>top_position</strong></p>
<blockquote>
<div><p>The index in the sequence that will end up at the top of the circle</p>
</div></blockquote>
<p><strong>feature_level_height</strong></p>
<blockquote>
<div><p>Width in inches of one “level” for feature arrows.</p>
</div></blockquote>
<p><strong>annotation_height</strong></p>
<blockquote>
<div><p>Width in inches of one “level” for feature annotations.</p>
</div></blockquote>
<p><strong>labels_spacing</strong></p>
<blockquote>
<div><p>Distance in basepairs to keep between labels to avoid “quasi-collisions”</p>
</div></blockquote>
<p><strong>**kw</strong></p>
<blockquote>
<div><p>Other keyword arguments - do not use, these parameters are allowed for
making it easier to use GraphicRecord and CircularGraphicRecord
interchangeably.</p>
</div></blockquote>
</dd>
</dl>
<dl class="method">
<dt id="dna_features_viewer.CircularGraphicRecord.coordinates_in_plot">
<code class="sig-name descname">coordinates_in_plot</code><span class="sig-paren">(</span><em class="sig-param">position</em>, <em class="sig-param">level</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/CircularGraphicRecord/CircularGraphicRecord.html#CircularGraphicRecord.coordinates_in_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.CircularGraphicRecord.coordinates_in_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a sequence position and height level to (x, y) coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.CircularGraphicRecord.determine_annotation_height">
<code class="sig-name descname">determine_annotation_height</code><span class="sig-paren">(</span><em class="sig-param">max_annotations_level</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/CircularGraphicRecord/CircularGraphicRecord.html#CircularGraphicRecord.determine_annotation_height"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.CircularGraphicRecord.determine_annotation_height" title="Permalink to this definition">¶</a></dt>
<dd><p>Auto-select the annotations height.</p>
<p>Annotation height is 0.2 at most, or else whatever will make
the figure a 5*radius tall rectangle where the circular plasmid
occupies the bottom-2 5th and the annotations occupy the top-3 5th.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.CircularGraphicRecord.finalize_ax">
<code class="sig-name descname">finalize_ax</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">features_levels</em>, <em class="sig-param">annotations_max_level</em>, <em class="sig-param">auto_figure_height=False</em>, <em class="sig-param">ideal_yspan=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/CircularGraphicRecord/CircularGraphicRecord.html#CircularGraphicRecord.finalize_ax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.CircularGraphicRecord.finalize_ax" title="Permalink to this definition">¶</a></dt>
<dd><p>Final display range and figure dimension tweakings.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.CircularGraphicRecord.initialize_ax">
<code class="sig-name descname">initialize_ax</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">draw_line</em>, <em class="sig-param">with_ruler</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/CircularGraphicRecord/CircularGraphicRecord.html#CircularGraphicRecord.initialize_ax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.CircularGraphicRecord.initialize_ax" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the ax with a circular line, sets limits, aspect etc.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.CircularGraphicRecord.plot_feature">
<code class="sig-name descname">plot_feature</code><span class="sig-paren">(</span><em class="sig-param">ax</em>, <em class="sig-param">feature</em>, <em class="sig-param">level</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/CircularGraphicRecord/CircularGraphicRecord.html#CircularGraphicRecord.plot_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.CircularGraphicRecord.plot_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot an ArrowWedge representing the feature at the giben height
level.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.CircularGraphicRecord.position_to_angle">
<code class="sig-name descname">position_to_angle</code><span class="sig-paren">(</span><em class="sig-param">position</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/CircularGraphicRecord/CircularGraphicRecord.html#CircularGraphicRecord.position_to_angle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.CircularGraphicRecord.position_to_angle" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a sequence position into an angle in the figure.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="graphicfeature">
<h2>GraphicFeature<a class="headerlink" href="#graphicfeature" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dna_features_viewer.GraphicFeature">
<em class="property">class </em><code class="sig-prename descclassname">dna_features_viewer.</code><code class="sig-name descname">GraphicFeature</code><span class="sig-paren">(</span><em class="sig-param">start=None</em>, <em class="sig-param">end=None</em>, <em class="sig-param">strand=None</em>, <em class="sig-param">label=None</em>, <em class="sig-param">color='#000080'</em>, <em class="sig-param">thickness=14</em>, <em class="sig-param">linewidth=1.0</em>, <em class="sig-param">linecolor='#000000'</em>, <em class="sig-param">fontdict=None</em>, <em class="sig-param">html=None</em>, <em class="sig-param">open_left=False</em>, <em class="sig-param">open_right=False</em>, <em class="sig-param">box_linewidth=1</em>, <em class="sig-param">box_color='auto'</em>, <em class="sig-param">**data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicFeature.html#GraphicFeature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicFeature" title="Permalink to this definition">¶</a></dt>
<dd><p>Genetic Feature to be plotted.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>start, end</strong></p>
<blockquote>
<div><p>Coordinates of the feature in the final sequence.</p>
</div></blockquote>
<p><strong>strand</strong></p>
<blockquote>
<div><p>Directionality of the feature. can be +1/-1/0 for direct sense,
anti-sense, or no directionality.</p>
</div></blockquote>
<p><strong>label</strong></p>
<blockquote>
<div><p>Short descriptive text associated and plotted with the feature</p>
</div></blockquote>
<p><strong>color</strong></p>
<blockquote>
<div><p>Color of the feature, any Matplotlib-compatible format is accepted,
such as “white”, “w”, “#ffffff”, (1,1,1), etc.</p>
</div></blockquote>
<p><strong>linecolor</strong></p>
<blockquote>
<div><p>Color of the feature’s border, any Matplotlib-compatible format is
accepted, such as “white”, “w”, “#ffffff”, (1,1,1), etc.</p>
</div></blockquote>
<p><strong>box_color</strong></p>
<blockquote>
<div><p>Color of the label box. Set to None for no box around the label.
Leave to “auto” for a box color that is a lightened version of the
feature’s color.</p>
</div></blockquote>
<p><strong>data</strong></p>
<blockquote>
<div><p>Any other keyword is kept into the feature.data[] dictionary.</p>
</div></blockquote>
<p><strong>fontdict</strong></p>
<blockquote>
<div><p>A Matplotlib fontdict for the font to be used in the label, e.g.
<code class="docutils literal notranslate"><span class="pre">size=11</span></code>, <code class="docutils literal notranslate"><span class="pre">weight='bold'</span></code>, <code class="docutils literal notranslate"><span class="pre">family='Helvetica'</span></code>, etc.</p>
</div></blockquote>
</dd>
</dl>
<dl class="method">
<dt id="dna_features_viewer.GraphicFeature.crop">
<code class="sig-name descname">crop</code><span class="sig-paren">(</span><em class="sig-param">window</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicFeature.html#GraphicFeature.crop"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicFeature.crop" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a the fragment of the feature that is in the window.</p>
<p>If there is no overlap between the feature location and the window,
None is returned.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicFeature.from_biopython_feature">
<em class="property">static </em><code class="sig-name descname">from_biopython_feature</code><span class="sig-paren">(</span><em class="sig-param">feature</em>, <em class="sig-param">**props</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicFeature.html#GraphicFeature.from_biopython_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicFeature.from_biopython_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a GraphicalFeature from a Biopython.Feature object.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicFeature.length">
<em class="property">property </em><code class="sig-name descname">length</code><a class="headerlink" href="#dna_features_viewer.GraphicFeature.length" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the feature (end-start)</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicFeature.overlaps_with">
<code class="sig-name descname">overlaps_with</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicFeature.html#GraphicFeature.overlaps_with"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicFeature.overlaps_with" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True iff the feature’s location overlaps with feature <cite>other</cite></p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicFeature.split_in_two">
<code class="sig-name descname">split_in_two</code><span class="sig-paren">(</span><em class="sig-param">x_coord=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/GraphicFeature.html#GraphicFeature.split_in_two"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.GraphicFeature.split_in_two" title="Permalink to this definition">¶</a></dt>
<dd><p>Return two features by cutting this feature at x_coord.</p>
</dd></dl>

<dl class="method">
<dt id="dna_features_viewer.GraphicFeature.x_center">
<em class="property">property </em><code class="sig-name descname">x_center</code><a class="headerlink" href="#dna_features_viewer.GraphicFeature.x_center" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the x-center of the feature, (start+end)/2</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-dna_features_viewer.biotools">
<span id="biotools"></span><h2>Biotools<a class="headerlink" href="#module-dna_features_viewer.biotools" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="dna_features_viewer.biotools.annotate_biopython_record">
<code class="sig-prename descclassname">dna_features_viewer.biotools.</code><code class="sig-name descname">annotate_biopython_record</code><span class="sig-paren">(</span><em class="sig-param">seqrecord</em>, <em class="sig-param">location='full'</em>, <em class="sig-param">feature_type='misc_feature'</em>, <em class="sig-param">margin=0</em>, <em class="sig-param">**qualifiers</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/biotools.html#annotate_biopython_record"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.biotools.annotate_biopython_record" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a feature to a Biopython SeqRecord.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seqrecord</strong></p>
<blockquote>
<div><p>The biopython seqrecord to be annotated.</p>
</div></blockquote>
<p><strong>location</strong></p>
<blockquote>
<div><p>Either (start, end) or (start, end, strand). (strand defaults to +1)</p>
</div></blockquote>
<p><strong>feature_type</strong></p>
<blockquote>
<div><p>The type associated with the feature</p>
</div></blockquote>
<p><strong>margin</strong></p>
<blockquote>
<div><p>Number of extra bases added on each side of the given location.</p>
</div></blockquote>
<p><strong>qualifiers</strong></p>
<blockquote>
<div><p>Dictionnary that will be the Biopython feature’s <cite>qualifiers</cite> attribute.</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dna_features_viewer.biotools.complement">
<code class="sig-prename descclassname">dna_features_viewer.biotools.</code><code class="sig-name descname">complement</code><span class="sig-paren">(</span><em class="sig-param">dna_sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/biotools.html#complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.biotools.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the complement of the DNA sequence.</p>
<p>For instance <code class="docutils literal notranslate"><span class="pre">complement(&quot;ATGCCG&quot;)</span></code> returns <code class="docutils literal notranslate"><span class="pre">&quot;TACGGC&quot;</span></code>.</p>
<p>Uses BioPython for speed.</p>
</dd></dl>

<dl class="function">
<dt id="dna_features_viewer.biotools.extract_graphical_translation">
<code class="sig-prename descclassname">dna_features_viewer.biotools.</code><code class="sig-name descname">extract_graphical_translation</code><span class="sig-paren">(</span><em class="sig-param">sequence</em>, <em class="sig-param">location</em>, <em class="sig-param">long_form=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/biotools.html#extract_graphical_translation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.biotools.extract_graphical_translation" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a string of the “graphical” translation of a sequence’s subsegment.</p>
<p>Here “graphical” means that the amino acid sequence is always given
left-to-right, as it will appear under the sequence in the plot. This matters
when the location is on the -1 strand. In this case, the amino-acids are
determined by (part of) the reverse-complement of the sequence, however
the sequence returned will be the mirror of the translated sequence, as
this is the left-to-right order in which the codons corresponding to the
amino-acids appear in the sequence.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong></p>
<blockquote>
<div><p>An “ATGC” string.</p>
</div></blockquote>
<p><strong>location</strong></p>
<blockquote>
<div><p>Either (start, end) or (start, end, strand), with strand in (0, 1, -1).</p>
</div></blockquote>
<p><strong>long_form</strong></p>
<blockquote>
<div><p>if True, a list of 3-letter amino acid representations is returned instead
([‘Ala’, ‘Ser’, …]).</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dna_features_viewer.biotools.reverse_complement">
<code class="sig-prename descclassname">dna_features_viewer.biotools.</code><code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/biotools.html#reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.biotools.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the reverse-complement of the DNA sequence.</p>
<p>For instance <code class="docutils literal notranslate"><span class="pre">complement(&quot;ATGCCG&quot;)</span></code> returns <code class="docutils literal notranslate"><span class="pre">&quot;GCCGTA&quot;</span></code>.</p>
<p>Uses BioPython for speed.</p>
</dd></dl>

<dl class="function">
<dt id="dna_features_viewer.biotools.translate">
<code class="sig-prename descclassname">dna_features_viewer.biotools.</code><code class="sig-name descname">translate</code><span class="sig-paren">(</span><em class="sig-param">dna_sequence</em>, <em class="sig-param">long_form=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dna_features_viewer/biotools.html#translate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dna_features_viewer.biotools.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate the DNA sequence into an amino-acids sequence MLKYQT…</p>
<p>If long_form is true, a list of 3-letter amino acid representations
is returned instead ([‘Ala’, ‘Ser’, …]).</p>
</dd></dl>

</div>
</div>


          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
  <li class="prev">
    <a href="index.html"
       title="previous chapter">← DNA Features Viewer</a>
  </li>
  <li class="next">
    <a href="tips.html"
       title="next chapter">Tips →</a>
  </li>
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2016, Edinburgh Genome Foundry.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0 with <a href="https://github.com/schettino72/sphinx_press_theme">Press Theme</a>.
</div>
            </div>
          </div>
      </page>
    </div>
    
    
  </body>
</html>